File: services/realtime-gateway/src/metrics.ts
import client from 'prom-client';

export const metricsRegistry = new client.Registry();

client.collectDefaultMetrics({
  register: metricsRegistry,
  prefix: 'hopdrop_realtime_gateway_'
});

export const httpRequestsTotal = new client.Counter({
  name: 'hopdrop_realtime_gateway_http_requests_total',
  help: 'Total HTTP requests processed by the realtime gateway',
  labelNames: ['method', 'route', 'status_code']
});

export const httpRequestDurationSeconds = new client.Histogram({
  name: 'hopdrop_realtime_gateway_http_request_duration_seconds',
  help: 'HTTP request duration in seconds for the realtime gateway',
  labelNames: ['method', 'route', 'status_code'],
  buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1]
});

export const activeSocketConnections = new client.Gauge({
  name: 'hopdrop_realtime_gateway_active_socket_connections',
  help: 'Active Socket.IO connections on the realtime gateway'
});

metricsRegistry.registerMetric(httpRequestsTotal);
metricsRegistry.registerMetric(httpRequestDurationSeconds);
metricsRegistry.registerMetric(activeSocketConnections);

export async function metricsPayload(): Promise<string> {
  return metricsRegistry.metrics();
}

export function recordHttpRequest(method: string, route: string, statusCode: number, durationSeconds: number): void {
  const labels = {
    method,
    route,
    status_code: String(statusCode)
  };

  httpRequestsTotal.inc(labels);
  httpRequestDurationSeconds.observe(labels, durationSeconds);
}

File: services/routing-search/app/metrics.py
from fastapi import Request
from prometheus_client import CONTENT_TYPE_LATEST, Counter, Histogram, generate_latest

http_requests_total = Counter(
    "hopdrop_routing_search_http_requests_total",
    "Total HTTP requests processed by routing-search",
    labelnames=("method", "route", "status_code"),
)

http_request_duration_seconds = Histogram(
    "hopdrop_routing_search_http_request_duration_seconds",
    "HTTP request duration in seconds for routing-search",
    labelnames=("method", "route", "status_code"),
    buckets=(0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2, 5),
)

http_errors_total = Counter(
    "hopdrop_routing_search_http_errors_total",
    "HTTP errors returned by routing-search",
    labelnames=("method", "route", "status_code"),
)


def metrics_payload() -> bytes:
    return generate_latest()


def metrics_content_type() -> str:
    return CONTENT_TYPE_LATEST


def get_metrics_route(request: Request) -> str:
    route = request.scope.get("route")
    route_path = getattr(route, "path", None)
    if route_path:
        return route_path
    return request.url.path


def record_http_request(request: Request, status_code: int, duration_seconds: float) -> None:
    route = get_metrics_route(request)
    labels = {
        "method": request.method,
        "route": route,
        "status_code": str(status_code),
    }

    http_requests_total.labels(**labels).inc()
    http_request_duration_seconds.labels(**labels).observe(duration_seconds)

    if status_code >= 500:
        http_errors_total.labels(**labels).inc()
